text/0043-disable-sbom.md

```diff
@@ -0,0 +1,50 @@
+# Provide Global Mechanism to Disable SBOM Generation
+
+## Summary
+
+Users should have a mechanism to disable the generation of SBOM documents
+during the build process.
+
+## Motivation
+
+The generation of the SBOM documents for an image can be very time-intensive in
+some cases. We have seen a doubling of build times for Node.js applications.
+Additionally, some users are not interested in taking advantage of the SBOM
+features of the buildpacks.
+
+## Detailed Explanation
+
+If a user wishes to produce an image that does not contain SBOM documents, they
+can set the `BP_GENERATE_SBOM` environment variable to `false`. When unset, the
+value of this environment variable is assumed to be `true`.
+
+When `BP_GENERATE_SBOM` is set to `false`, none of the buildpacks executing
+during the build phase should generate or attach an SBOM in their outputs. This
+would apply to both new (Syft, CycloneDX, and SPDX formats) and old (label) SBOM
+outputs.
+
+## Rationale and Alternatives
+
+Alternatively, we could not provide an mechanism to disable SBOM generation.
+This does not seem like a great option as the build-time penalty introduced by
+SBOM generation can be substantial.
```

Comment:
Does anyone feel strongly enough that everyone should always have an SBOM included? SBOM is one of the marquee features of buildpacks. I assume that there are/will be tools built that consume the SBOM and do interesting stuff with it. Not having an SBOM could impact the success/adoption of those tools.

Do we feel like the performance hit here is related to the tool we are using? Is there a chance this tooling will get better/faster with some time/improvements?

Once we enable a feature like this (disabling SBOM), it would be hard to take it back.

I'm not sure exactly how I feel about the above statements, but I wanted to voice some contrary talking points to see what folks think.

Reply:
I also wonder about how this might impact adoption for large companies. If you allow disabling SBOM at the developer level, there isn't really a way to ensure that SBOM is there like if that is a company security/compliance requirement. 

If a dev can go in and turn off SBOM, they could bypass that requirement. Then additional work needs to be done to catch and force users to generate SBOM.

Reply:
I think optionality will ensure that you don't discourage casual developers from adopting buildpacks due to much longer build times due to sbom. At the enterprise level, lots of things can be disabled by a developer choosing to avoid security practices set in place by their leadership or security team.

Reply:
>I think optionality will ensure that you don't discourage casual developers from adopting buildpacks due to much longer build times due to sbom.

Right, but it's going to default to being enabled. It'll be slow out of the box. That means a casual developer needs to know to disable this to speed things up. I'm sure some will, but some won't too.

I definitely don't think we should just expect users to be Ok with it being slower, but perhaps there are other options. Improvements to the scanning tools? Different scanning tools? Performing a "lite" scan vs a full scan (possibly we could adjust the options to change the fidelity of information generated to speed up the scan)? Buildpack dev/test/debug mode, where we build the app image in a way that is only suitable for development but in that case, we don't scan (perhaps this would be a deterrent to using this as a way to bypass SBOM generation)?

I'm just concerned about making this optional. SBOM is an important feature of buildpacks. It would be nice if Paketo as a project can declare that all of our buildpacks generate SBOM. If we make it optional, that statement is diluted.

Reply:
@fg-j I think in my mind SBOM is a security-related component. Do we have prior art to allow users to disable security-related features?

Reply:
Do you think source code removal falls into that category, @dmikusa-pivotal? In that case, the ability to switch off source code removal feels like a comparable example.

Reply:
Yes, I could see arguments for that being security-related. 

I was thinking about this from the Java perspective. What I found that I think matches up is the ability to force a JDK to be installed in the runtime container. That would be prior art for something "security-related" (that you can opt-out of and it makes the images less secure). That's widely considered bad practice, cause it's including compiler tools in your runtime image. We allow that option though because there are some use cases that require it (they do on-the-fly compilation).

So it seems like there are a couple of things.

Ultimately what I think would negate this argument, but I'm not sure is possible, would be if a setting like this could be controlled at the builder level or across a platform. In that way, users can set it one way for local development but platform operators/companies can dictate a policy that cannot be side-stepped for production environments. I think there is going to be a desire for control over SBOM generation at the Ops level and that it'll be policy requirements for companies. If we can't deliver that, I worry it could impact adoption.

Reply:
I think that type of platform-wide override would need to be something that lived at the platform level. For instance, could there not be a `kpack` feature that enforced a specific environment variable was always set to a specific value, no matter what a developer might have provided. The interface for either of these features would still be an environment variable, but its up to the platform to enable this platform-wide override functionality.

Reply:
No disagreements there, a platform implementation makes sense. There is no present implementation though, so if we allow this to be disabled it could be a blocker for adoption with some users.

I'd be more OK with this if the RFC did not require language families to implement it. There is sufficient prior art that settings like this can be adoption at the project level and made optional, like the [RFC 32](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0032-reloadable-process-types.md) and [RFC 37](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0037-remote-debug.md). They define the interface, but don't require it to be implemented.